### PR TITLE
Provide Persistance for Persistance.Reference

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Provide default `Persistance` implementation for `Persistance.Reference` to allow us to persist fields of that type easily. The reference is used to _load object lazily_. When constructing the objects in memory one uses `Reference.of(anObject)`, stores it in a field and after deserializing one gets an instance of the `Reference` back (but without the original object being deserialized). Later one calls `ref.get(TypeOfTheObject.class)` to deserialize also instance of the original object.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      style guides. 
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

File: engine/runtime-parser/src/test/java/org/enso/compiler/core/IrPersistanceTest.java [hunk 1]
```diff
@@ -470,4 +501,11 @@ public record ServiceSupply(Service supply) {}
 
   @Persistable(clazz = IdHolder.class, id = 432876)
   public record IdHolder(UUID id) {}
+
+  @Persistable(clazz = RefHolder.class, id = 436872)
+  public record RefHolder(Persistance.Reference<UUID> id) {
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/test/java/org/enso/compiler/core/IrPersistanceTest.java:506]: Have a class with field `Persistance.Reference` and write/generate `Persistance` instance for it. In this case we generate the class with `@Persistable(clazz = RefHolder.class...` annotation.

File: engine/runtime-parser/src/test/java/org/enso/compiler/core/IrPersistanceTest.java [hunk 2]
```diff
@@ -68,7 +69,37 @@ public void identifiedLocationNoUUID() throws Exception {
   }
 
   @Test
-  public void idHolderNoUUID() throws Exception {
+  public void refHolderWithUUID() throws Exception {
+    var id = UUID.randomUUID();
+    var il = new RefHolder(id);
+    var in = serde(RefHolder.class, il, -1, null);
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/test/java/org/enso/compiler/core/IrPersistanceTest.java:75]: This stores an instance of `RefHolder` and reads another instace of `RefHolder` back from the array. Only then when one `in.id().get(UUID.class)` the instance of `UUID` gets deserialized.

**Review by hubertp — APPROVED** [review]

**Comment by radeusgd** [comment]
As discussed, here's the repro for the problem with cycles in serialization that I'm encountering:

- Check out the branch: https://github.com/enso-org/enso/tree/wip/radeusgd/cycle-in-serialization
   - I added there the mechanism for detecting the cycles (now it works, when we were chatting I had a missing case where the objects need to be registered). You can revert 47311b0a if you prefer to not have it.
- Just running `buildEngineDistribution` triggers the issue.
- Alternatively, I prefer to `buildEngineDistributionNoIndex` to build the artefacts, followed by manually running `.\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\bin\enso --no-global-cache --no-compile-dependencies --compile .\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\AWS\0.0.0-dev\` to trigger compilation of the `AWS` library (it's a small one so I figured out it will be easier for debugging, but any library incl. `Base` should work) - this way it is easier for me to attach the debugger.

image

After some digging and relying on the cycle detector, it our initial hypothesis (`Name.Literal(Text, ...)` being teh culprit) seems to be wrong and after all the object that is **really** causing the cycle seems to be
```
Cons(Operation,List(Argument(kind,false,Some(org.enso.persist.PerMemoryReference@2389e02f)), Argument(expressions,false,Some(org.enso.persist.PerMemoryReference@6db609e7)), Argument(metadata,true,Some(org.enso.persist.PerMemoryReference@7e5bfb5a))))
```
which is quite expected since it is indeed the `Cons`/`Argument` pair that I have been modifying - adding the third `Option[Reference[Expression]]` to it. Indeed removing the expression from there was fixing the issue.

> Overall the cycle seems to be `Cons -> Argument -> Expression (Application.Prefix) -> Name.Literal -> MetadataStorage(GlobalNames) -> BindingsMap.Resolution -> ResolvedType -> Type -> Cons`

**Comment by radeusgd** [comment]
It seems that I've got a _(partial)_ fix for the issue. It seems that writing is fixed by ensuring that the currently written object is registered _before_ serializing it (so that once we encounter the cycle during serialization, we already see the backreference and avoid the infinite loop).

Seems to be taken care of by this patch:
```diff
diff --git a/lib/java/persistance/src/main/java/org/enso/persist/PerGenerator.java b/lib/java/persistance/src/main/java/org/enso/persist/PerGenerator.java
--- a/lib/java/persistance/src/main/java/org/enso/persist/PerGenerator.java	(revision c1fead2ffa276279c28447db7e17118cc92cb2ed)
+++ b/lib/java/persistance/src/main/java/org/enso/persist/PerGenerator.java	(revision 967cffa4b916e42f194e565f4e664fb7d293c1cd)
@@ -59,14 +59,15 @@
     java.lang.Object obj = writeReplace.apply(t);
     java.lang.Integer found = knownObjects.get(obj);
     if (found == null) {
+      found = this.position;
+      knownObjects.put(obj, found);
+
       org.enso.persist.Persistance<?> p = map.forType(obj.getClass());
       java.io.ByteArrayOutputStream os = new ByteArrayOutputStream();
       p.writeInline(obj, new ReferenceOutput(this, os));
-      found = this.position;
       byte[] arr = os.toByteArray();
       main.write(arr);
       this.position += arr.length;
-      knownObjects.put(obj, found);
     }
     return found;
   }
@@ -83,18 +84,18 @@
     }
     java.lang.Integer found = knownObjects.get(obj);
     if (found == null) {
+      found = position;
+      knownObjects.put(obj, found);
 
       var os = new ByteArrayOutputStream();
       var osData = new ReferenceOutput(this, os);
       p.writeInline(obj, osData);
-      found = position;
       if (os.size() == 0) {
         os.write(0);
       }
       byte[] arr = os.toByteArray();
       main.write(arr);
       position += arr.length;
-      knownObjects.put(obj, found);
       if (histogram != null) {
         histogram.register(obj.getClass(), arr.length);
       }
```

After this patch I was able to rebuild indices.

I'm slightly worried that there may be some issues when reading this back, since I'm getting errors like:
```
[info] Generating index for built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\AWS
[WARN] [2024-03-08T13:30:39+01:00] [enso.org.enso.interpreter.caches.Cache] `Standard.AWS` in built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\AWS\0.0.0-dev\.enso\cache\bindings\0.0.0-dev\Standard\AWS.bindings failed to load: No persistance for 1952542308
[info] Generating index for built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Base
[info] Generating index for built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Database
[WARN] [2024-03-08T13:30:56+01:00] [enso.org.enso.interpreter.caches.Cache] `Standard.Base` in X:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Base\0.0.0-dev\.enso\cache\bindings\0.0.0-dev\Standard\Base.bindings failed to load: No persistance for 279137
```

But I'm not 100% sure if this is it or just unrelated errors?

More problems after clean build:
```
[info] Generating index for built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\AWS
[info] Generating index for built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Base
[WARN] [2024-03-08T13:37:34+01:00] [enso.org.enso.interpreter.caches.Cache] `Standard.Builtins.Main` in C:\Users\DELL\AppData\Local\enso\cache\ir\Standard\Builtins\0.0.0-dev\0.0.0-dev\Standard\Builtins\Main.ir failed to load: No persistance for 541301
[info] Generating index for built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Database
[WARN] [2024-03-08T13:37:44+01:00] [enso.org.enso.interpreter.caches.Cache] `Standard.Builtins.Main` in C:\Users\DELL\AppData\Local\enso\cache\ir\Standard\Builtins\0.0.0-dev\0.0.0-dev\Standard\Builtins\Main.ir failed to load: No persistance for 541301
[WARN] [2024-03-08T13:37:45+01:00] [enso.org.enso.interpreter.caches.Cache] `Standard.Base` in X:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Base\0.0.0-dev\.enso\cache\bindings\0.0.0-dev\Standard\Base.bindings failed to load: No persistance for 279137
[info] Generating index for built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Examples
[WARN] [2024-03-08T13:37:55+01:00] [enso.org.enso.interpreter.caches.Cache] `Standard.Builtins.Main` in C:\Users\DELL\AppData\Local\enso\cache\ir\Standard\Builtins\0.0.0-dev\0.0.0-dev\Standard\Builtins\Main.ir failed to load: No persistance for 541301
[info] Generating index for built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Geo
[WARN] [2024-03-08T13:38:01+01:00] [enso.org.enso.interpreter.caches.Cache] `Standard.Builtins.Main` in C:\Users\DELL\AppData\Local\enso\cache\ir\Standard\Builtins\0.0.0-dev\0.0.0-dev\Standard\Builtins\Main.ir failed to load: No persistance for 541301
[info] Generating index for built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Google_Api
[WARN] [2024-03-08T13:38:07+01:00] [enso.org.enso.interpreter.caches.Cache] `Standard.Builtins.Main` in C:\Users\DELL\AppData\Local\enso\cache\ir\Standard\Builtins\0.0.0-dev\0.0.0-dev\Standard\Builtins\Main.ir failed to load: No persistance for 541301
[info] Generating index for built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Image
[WARN] [2024-03-08T13:38:13+01:00] [enso.org.enso.interpreter.caches.Cache] `Standard.Builtins.Main` in C:\Users\DELL\AppData\Local\enso\cache\ir\Standard\Builtins\0.0.0-dev\0.0.0-dev\Standard\Builtins\Main.ir failed to load: No persistance for 541301
[info] Generating index for built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Searcher
[WARN] [2024-03-08T13:38:18+01:00] [enso.org.enso.interpreter.caches.Cache] `Standard.Builtins.Main` in C:\Users\DELL\AppData\Local\enso\cache\ir\Standard\Builtins\0.0.0-dev\0.0.0-dev\Standard\Builtins\Main.ir failed to load: No persistance for 541301
[info] Generating index for built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Table
[WARN] [2024-03-08T13:38:23+01:00] [enso.org.enso.interpreter.caches.Cache] `Standard.Builtins.Main` in C:\Users\DELL\AppData\Local\enso\cache\ir\Standard\Builtins\0.0.0-dev\0.0.0-dev\Standard\Builtins\Main.ir failed to load: No persistance for 541301
[info] Generating index for built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test
[WARN] [2024-03-08T13:38:32+01:00] [enso.org.enso.interpreter.caches.Cache] `Standard.Builtins.Main` in C:\Users\DELL\AppData\Local\enso\cache\ir\Standard\Builtins\0.0.0-dev\0.0.0-dev\Standard\Builtins\Main.ir failed to load: No persistance for 541301
[info] Generating index for built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Visualization
[WARN] [2024-03-08T13:38:39+01:00] [enso.org.enso.interpreter.caches.Cache] `Standard.Builtins.Main` in C:\Users\DELL\AppData\Local\enso\cache\ir\Standard\Builtins\0.0.0-dev\0.0.0-dev\Standard\Builtins\Main.ir failed to load: No persistance for 541301
[WARN] [2024-03-08T13:38:39+01:00] [enso.org.enso.interpreter.caches.Cache] `Standard.Table` in X:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Table\0.0.0-dev\.enso\cache\bindings\0.0.0-dev\Standard\Table.bindings failed to load: No persistance for 1634467845
[WARN] [2024-03-08T13:38:40+01:00] [enso.org.enso.interpreter.caches.Cache] `Standard.Database` in X:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Database\0.0.0-dev\.enso\cache\bindings\0.0.0-dev\Standard\Database.bindings failed to load: No persistance for 1953066862
[info] Engine package created at built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev
```

**Comment by JaroslavTulach** [comment]
> It seems that I've got a _(partial)_ fix for the issue. 

I have a feeling it fails a lot of `IrPersistanceTest`s. I'll integrate this PR without further diffs. Please report another issue requesting ability to store cycles in the graph. Extending `IrPersistanceTest` or some other to demonstrate the problem would be welcomed. Not completely necessary, I guess I understand now where the problem is and how the fix should look like.

**Comment by JaroslavTulach** [comment]
Request to also support "cycles" in the graph is 
- #9361

**Pull request merged**